Tools/unix/cpmtools/mkfs.cpm.1
.TH MKFS.CPM 1 "October 25, 2014" "CP/M tools" "User commands"
.SH NAME \"{{{roff}}}\"{{{
mkfs.cpm \- make a CP/M file system
.\"}}}
.SH SYNOPSIS \"{{{
.ad l
.B mkfs.cpm
.RB [ \-f
.IR format ]
.RB [ \-b
.IR boot ]
.RB [ \-L
.IR label ]
.RB [ \-t ]
.I image
.ad b
.\"}}}
.SH DESCRIPTION \"{{{
\fBmkfs.cpm\fP makes a CP/M file system on an image file or device.
.\"}}}
.SH OPTIONS \"{{{
.IP "\fB\-f\fP \fIformat\fP"
Use the given CP/M disk \fIformat\fP instead of the default format.
.IP "\fB\-b\fP \fIbootblock\fP"
Write the contents of the file \fIbootblock\fP to the system tracks
instead of filling them with 0xe5.  This option can be used up to four
times.  The file contents (typically boot block, CCP, BDOS and BIOS)
are written to sequential sectors, padding with 0xe5 if needed.
.IP "\fB\-L\fP \fIlabel\fP"
Label the file system.  This is only supported by CP/M Plus.
.IP "\fB\-t\fP"
Create time stamps.
.\"}}}
.SH "RETURN VALUE" \"{{{
Upon successful completion, exit code 0 is returned.
.\"}}}
.SH ERRORS \"{{{
Any errors are indicated by exit code 1.
.\"}}}
.SH ENVIRONMENT \"{{{
CPMTOOLSFMT     Default format
.\"}}}
.SH FILES \"{{{
${prefix}/share/diskdefs	CP/M disk format definitions
.\"}}}
.SH AUTHORS \"{{{
This program is copyright 1997\(en2012 Michael Haardt
<michael@moria.de>.  The Windows port is copyright 2000, 2001, 2011 John Elliott
<jce@seasip.demon.co.uk>.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program.  If not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"}}}
.SH "SEE ALSO" \"{{{
.IR fsck.cpm (1),
.IR cpmls (1),
.IR mkfs (1),
.IR cpm (5)
.\"}}}


branches/dwg/LSource/src/cpmtools-2.13/mkfs.cpm.c
/* #includes */ /*{{{C}}}*//*{{{*/
#include "config.h"

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "getopt_.h"
#include "cpmfs.h"

#ifdef USE_DMALLOC
#include <dmalloc.h>
#endif
/*}}}*/
/* #defines */ /*{{{*/
#ifndef O_BINARY
#define O_BINARY 0
#endif
/*}}}*/

/* mkfs -- make file system */ /*{{{*/
static int mkfs(struct cpmSuperBlock *drive, const char *name, const char *label, char *bootTracks)
{
  /* variables */ /*{{{*/
  int i;
  char buf[128];
  char firstbuf[128];
  int fd;
  int bytes;
  int trkbytes;
  /*}}}*/

  /* open image file */ /*{{{*/
  if ((fd = open(name, O_BINARY|O_CREAT|O_WRONLY, 0666)) < 0)
  {
    boo=strerror(errno);
    return -1;
  }
  /*}}}*/
  /* write system tracks */ /*{{{*/
  /* this initialises only whole tracks, so it skew is not an issue */
  trkbytes=drive->secLength*drive->sectrk;
  for (i=0; i<trkbytes*drive->boottrk; i+=drive->secLength) if (write(fd, bootTracks+i, drive->secLength)!=drive->secLength)
  {
    boo=strerror(errno);
    close(fd);
    return -1;
  }
  /*}}}*/
  /* write directory */ /*{{{*/
  memset(buf,0xe5,128);
  bytes=drive->maxdir*32;
  if (bytes%trkbytes) bytes=((bytes+trkbytes)/trkbytes)*trkbytes;
  if (drive->type==CPMFS_P2DOS || drive->type==CPMFS_DR3) buf[3*32]=0x21;
  memcpy(firstbuf,buf,128);
  if (drive->type==CPMFS_DR3)
  {
    time_t now;
    struct tm *t;
    int min,hour,days;

    firstbuf[0]=0x20;
    for (i=0; i<11 && *label; ++i,++label) firstbuf[1+i]=toupper(*label&0x7f);
    while (i<11) firstbuf[1+i++]=' ';
    firstbuf[12]=0x11; /* label set and first time stamp is creation date */
    memset(&firstbuf[13],0,1+2+8);
    time(&now);
    t=localtime(&now);
    min=((t->tm_min/10)<<4)|(t->tm_min%10);
    hour=((t->tm_hour/10)<<4)|(t->tm_hour%10);
    for (i=1978,days=0; i < 1900 + t->tm_year; ++i)
    {
      days+=365;
      if (i%4==0 && (i%100!=0 || i%400==0)) ++days;
    }
    days += t->tm_yday + 1;
    firstbuf[24]=firstbuf[28]=days&0xff; firstbuf[25]=firstbuf[29]=days>>8;
    firstbuf[26]=firstbuf[30]=hour;
    firstbuf[27]=firstbuf[31]=min;
  }
  for (i=0; i < bytes; i += 128) if (write(fd, i==0 ? firstbuf : buf, 128)!=128)
  {
    boo=strerror(errno);
    close(fd);
    return -1;
  }
  /*}}}*/
  /* close image file */ /*{{{*/
  if (close(fd)==-1)
  {
    boo=strerror(errno);
    return -1;
  }
  /*}}}*/
  return 0;
}
/*}}}*/

const char cmd[]="mkfs.cpm";

int main(int argc, char *argv[]) /*{{{*/
{
  char *image;
  const char *format=FORMAT;
  int c,usage=0;
  struct cpmSuperBlock drive;
  struct cpmInode root;
  const char *label="unlabeled";
  size_t bootTrackSize,used;
  char *bootTracks;
  const char *boot[4]={(const char*)0,(const char*)0,(const char*)0,(const char*)0};

  while ((c=getopt(argc,argv,"b:f:L:h?"))!=EOF) switch(c)
  {
    case 'b':
    {
      if (boot[0]==(const char*)0) boot[0]=optarg;
      else if (boot[1]==(const char*)0) boot[1]=optarg;
      else if (boot[2]==(const char*)0) boot[2]=optarg;
      else if (boot[3]==(const char*)0) boot[3]=optarg;
      else usage=1;
      break;
    }
    case 'f': format=optarg; break;
    case 'L': label=optarg; break;
    case 'h':
    case '?': usage=1; break;
  }

  if (optind!=(argc-1)) usage=1;
  else image=argv[optind++];

  if (usage)
  {
    fprintf(stderr,"Usage: %s [-f format] [-b boot] [-L label] image\n",cmd);
    exit(1);
  }
  drive.dev.opened=0;
  cpmReadSuper(&drive,&root,format);
  bootTrackSize=drive.boottrk*drive.secLength*drive.sectrk;
  if ((bootTracks=malloc(bootTrackSize))==(void*)0)
  {
    fprintf(stderr,"%s: can not allocate boot track buffer: %s\n",cmd,strerror(errno));
    exit(1);
  }
  memset(bootTracks,0xe5,bootTrackSize);
  used=0; 
  for (c=0; c<4 && boot[c]; ++c)
  {
    int fd;
    size_t size;

    if ((fd=open(boot[c],O_BINARY|O_RDONLY))==-1)
    {
      fprintf(stderr,"%s: can not open %s: %s\n",cmd,boot[c],strerror(errno));
      exit(1);
    }
    size=read(fd,bootTracks+used,bootTrackSize-used);
#if 0
    fprintf(stderr,"%d %04x %s\n",c,used+0x800,boot[c]);
#endif
    if (size%drive.secLength) size=(size|(drive.secLength-1))+1;
    used+=size;
    close(fd);
  }
  if (mkfs(&drive,image,label,bootTracks)==-1)
  {
    fprintf(stderr,"%s: can not make new file system: %s\n",cmd,boo);
    exit(1);
  }
  else exit(0);
}
/*}}}*/
